pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_index.go

```diff
@@ -0,0 +1,438 @@
+	elms ElementResultSet,
+	sie *scpb.SecondaryIndex,
+	referencedColumnIDs []descpb.ColumnID,
+) bool {
+	result := false
+
+	// Check all indexes (both primary and secondary) to see if we can find a replacement candidate.
+	elms.ForEachElementStatus(func(current scpb.Status, target scpb.TargetStatus, e scpb.Element) {
+		if result {
+			return
+		}
+
+		switch t := e.(type) {
+		case *scpb.SecondaryIndex:
+			if t.IndexID != sie.IndexID && isValidReferencedUniqueConstraint(b, &t.Index, referencedColumnIDs) {
+				result = true
+			}
+		case *scpb.PrimaryIndex:
+			if isValidReferencedUniqueConstraint(b, &t.Index, referencedColumnIDs) {
+				result = true
+			}
+		}
+	})
+
+	return result
+}
+
+// Attempt to drop all views that depend on the to be dropped index if CASCADE.
+// Panic if there is a dependent view but drop behavior is not CASCADE.
+func mayDropDependentViews(
+	b BuildCtx, toBeDroppedSecondaryIndexElms ElementResultSet, dropBehavior tree.DropBehavior,
+) {
+	_, _, sie := scpb.FindSecondaryIndex(toBeDroppedSecondaryIndexElms)
+	_, _, sine := scpb.FindIndexName(toBeDroppedSecondaryIndexElms)
+	if sie == nil || sine == nil {
+		panic("invalid toBeDroppedSecondaryIndexElms; does not contain a secondary index element.")
+	}
+
+	scpb.ForEachView(b.BackReferences(sie.TableID), func(current scpb.Status, target scpb.TargetStatus, ve *scpb.View) {
+		for _, forwardRef := range ve.ForwardReferences {
+			if forwardRef.IndexID == sie.IndexID {
+				// This view depends on the to-be-dropped index;
+				if dropBehavior != tree.DropCascade {
+					// Get view name for panic message
+					_, _, ns := scpb.FindNamespace(b.QueryByID(ve.ViewID))
+					panic(errors.WithHintf(
+						sqlerrors.NewDependentObjectErrorf("cannot drop index %q because view %q depends on it",
+							sine.Name, ns.Name), "you can drop %q instead.", ns.Name))
+				} else {
+					dropCascadeDescriptor(b, ve.ViewID)
+				}
+			}
+		}
+	})
+}
+
+// Attempt to drop all FK constraints that depend on the to be dropped index if CASCADE.
+// A FK constraint can only exist if there is `PRIMARY KEY` or `UNIQUE` constraint on the referenced columns
+// in the child table.
+// This is relevant if we're dropping a unique index whose `UNIQUE` constraints serves some FK constraints from
+// other tables. In this case, we attempt to find a replacement constraint to serve this FK constraint. If we can,
+// then we can proceed to drop the index. Otherwise, we will need to drop the FK constraint as well (if CASCADE of
+// course).
+// Panic if there is a dependent view but drop behavior is not CASCADE.
+func mayDropDependentFKConstraints(
+	b BuildCtx, toBeDroppedSecondaryIndexElms ElementResultSet, dropBehavior tree.DropBehavior,
+) {
+	_, _, sie := scpb.FindSecondaryIndex(toBeDroppedSecondaryIndexElms)
+	_, _, sine := scpb.FindIndexName(toBeDroppedSecondaryIndexElms)
+	if sie == nil || sine == nil {
+		panic("invalid toBeDroppedSecondaryIndexElms; does not contain a secondary index element.")
```

Comment:
make this an error, and if it's not an assertion failure, then properly classify it.